shared_python/ValidateTags.py

```diff
@@ -0,0 +1,187 @@
+from logging import Logger
+
+class ValidateTags(object):
+    def __init__(self, log):
+        self.log = log
+
+    ## Dictionaries for tag validation
+    dict_tag_type = {
+        "rating": 1,
+        "warnings": 2,
+        "categories": 3,
+        "fandoms": 4,
+        "characters": 5,
+        "relationships": 6,
+        "tags": 7
+    }
+    dict_rating = {
+        "Not Rated": 1,
+        "General Audiences": 2,
+        "Teen And Up Audiences": 3,
+        "Mature": 4,
+        "Explicit": 5
+    }
+    dict_warning = {
+        "Choose Not To Use Archive Warnings": 1,
+        "Graphic Depictions Of Violence": 2,
+        "Major Character Death": 3,
+        "No Archive Warnings Apply": 4,
+        "Rape/Non-Con": 5,
+        "Underage": 6
+    }
+    dict_category = {
+        "Gen": 1,
+        "F/M": 2,
+        "M/M": 3,
+        "F/F": 4,
+        "Multi": 5,
+        "Other": 6
+    }
+
+    ## Functions for accessing dictionaries
+    def identify_tag_type(self, tag_type):
+        return self.dict_tag_type.get(tag_type, 0)
+
+    def identify_rating(self, tag):
+        return self.dict_rating.get(tag, 0)
+
+    def identify_warning(self, tag):
+        return self.dict_warning.get(tag, 0)
+
+    def identify_category(self, tag):
+        return self.dict_category.get(tag, 0)
+
+    ## Print statements
+    def print_tag_correction(self, before, after):
+        print('\r\033[96m', end='')
+        self.log.info('Correction successful. "' + before + '" is now "' + after+ '"')
+        print('\x1b[0m', end='')
+
+    def print_tag_warning(self, tag, isType):
+        print('\r\033[93m', end='')
+        if (isType):
+            self.log.warning('Warning: "' + tag + '" is not a valid TAG TYPE. Attempting self correction...')
+        else:
+            self.log.warning('Warning: "' + tag + '" is not a valid TAG. Attempting self correction...')
```

Comment:
Looking at the screenshots, I was thinking an enhancement here would be to say what type of tag it isn't valid for. `Teen & up audiences` is technically a valid "tag" but it is specifically an invalid (AO3) `rating`. But it can be future enhancement if that's difficult to sort out at this point.

Reply:
This is a good point. I had thought about changing the warning to list the tag_type, but I got lazy and didn't want to add more parameters. I've added it in now, so it will describe the tag type it's invalid for.
![image](https://user-images.githubusercontent.com/65837245/137646458-0718f420-e7ab-417f-9231-d172d8b35329.png)

Reply:
That's excellent, thanks! 🎉